File: libro/132_factorial_robusto.cpp
/* ********************************************************
* Programa: FactorialRobusto
*
* Descripción:
*   Este programa calcula el factorial de los primeros
*   números naturales, de forma recursiva
******************************************************** */

#include <stdio.h>
#include <limits.h>

int FactorialRobusto( int n ) {
  int f = 1;

  if (n < 0){
    throw 0;
  }

  for (int k = 2; k <= n; k++) {
    if (f > INT_MAX/k){
      throw k;
    }
    f = f * k;
  }
  return f;
}

void EscribirFactorial(int num){
  try{
    printf("%2d! vale:%10d\n", num, FactorialRobusto(num));
  }
  catch (int e){
    printf("%2d! excepción: ", num);
    if (e == 0){
      printf("Factorial de número negativo\n");
    }else{
      printf("Superado límite al evaluar %2d!\n", num);
    }
  }
}

int main() {

  for ( int i = 5; i >= - 2; i--) {
    EscribirFactorial( i );
  }
  printf( "\n" );
  for (int i = 6; i <= 15; i++) {
    EscribirFactorial( i );
  }
}

File: libro/132_factorial_robusto.py
# Programa: Fatorial Robusto
# Descripción:
#   Este programa es una calculadora que suma,
#   resta, multiplica y divide fracciones
import sys

class NumeroException(BaseException):
    def __init__(self, numero):
        self.numero = numero

# Procedimiento para simplificar la fracción n/d
def FactorialRobusto(n):

    f = 1

    if n<0:
        raise NumeroException(numero=0)

    for k in range(2, n+1):
        if f>sys.maxsize:
            raise NumeroException(numero=f)
        f = f*k

    return f


def EscribirFactorial(num):

  try:
    fact = FactorialRobusto(num)
    print("{0:2d}! vale:{1:10d}".format(num, fact))
  
  except NumeroException as e:
    if e.numero == 0:
      print("{0:2d}! excepción: Factorial de número negativo".format(num))
    else:
      print("{0:2d}! excepción: Superado límite al evaluar {1:2d}!".format(num, num))


# Programa principal
if __name__ == "__main__":

    for i in range(5, -3, -1):
        EscribirFactorial(i)
  
    for i in range(6, 16):
        EscribirFactorial(i)